src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -1608,6 +1615,71 @@ abstract class ManagedStrategy(
         }
     }
 
+    private fun methodAnalysisSectionType(
+        owner: Any?,
+        className: String,
+        methodName: String,
+        atomicMethodDescriptor: AtomicMethodDescriptor?,
+        deterministicMethodDescriptor: DeterministicMethodDescriptor<*, *>?,
+    ): AnalysisSectionType {
+        val ownerName = owner?.javaClass?.canonicalName ?: className
+        if (atomicMethodDescriptor != null) {
+            return AnalysisSectionType.ATOMIC
+        }
+        // TODO: decide if we need to introduce special `DETERMINISTIC` guarantee?
+        if (deterministicMethodDescriptor != null) {
+            return AnalysisSectionType.IGNORE
+        }
+        // Ignore methods called on standard I/O streams
+        when (owner) {
+            System.`in`, System.out, System.err -> return AnalysisSectionType.IGNORE
+        }
+        val silentSection = getDefaultSilentSectionType(ownerName, methodName)
+            ?.ensure { it.isSilent() }
+        if (silentSection != null) {
+            return silentSection
+        }
+        userDefinedGuarantees?.forEach { guarantee ->
+            if (guarantee.classPredicate(ownerName) && guarantee.methodPredicate(methodName)) {
+                return guarantee.type
+            }
+        }
+        return AnalysisSectionType.NORMAL
+    }
+
+    private fun enterAnalysisSection(threadId: ThreadId, section: AnalysisSectionType) {
+        val analysisSectionStack = this.analysisSectionStack[threadId]!!
+        val currentSection = analysisSectionStack.lastOrNull()
+        if (currentSection != null && currentSection.isCallStackPropagating() && section < currentSection) {
+            analysisSectionStack.add(currentSection)
+        } else {
+            analysisSectionStack.add(section)
+        }
+        if (section == AnalysisSectionType.IGNORE ||
+            // TODO: atomic should have different semantics compared to ignored
+            section == AnalysisSectionType.ATOMIC
+        ) {
+            enterIgnoredSection()
+        }
+    }
+
+    private fun leaveAnalysisSection(threadId: ThreadId, section: AnalysisSectionType) {
+        if (section == AnalysisSectionType.IGNORE ||
+            // TODO: atomic should have different semantics compared to ignored
```

Comment:
What do you mean?

Reply:
Before this PR, `ATOMIC` and `IGNORE` sections were handled similarly, by just entering ingored section and disabling the analysis completely. 

However, I think `ATOMIC` should be handled different. It should be similar to `SILENT` (so analysis is still run inside it, but with no switch or trace points created). But unlike `SILENT` section, in `ATOMIC` section there should be no thread switches at all. 

If some thread switch is enforced inside `ATOMIC` section because of blocking, it means that the method is not truly atomic, and shouldn't be marked as `ATOMIC`.